sql/plan/insert.go

```diff
@@ -119,17 +137,37 @@ func (p *InsertInto) Execute(ctx *sql.Context) (int, error) {
 		if err == io.EOF {
 			break
 		}
-
 		if err != nil {
 			_ = iter.Close()
 			return i, err
 		}
 
-		if err := insertable.Insert(ctx, row); err != nil {
+		err = p.validateNullability(ctx, dstSchema, row)
+		if err != nil {
 			_ = iter.Close()
 			return i, err
 		}
 
+		if replaceable != nil {
+			if err = replaceable.Delete(ctx, row); err != nil {
+				if err != sql.ErrDeleteRowNotFound {
+					_ = iter.Close()
+					return i, err
+				}
+			} else {
+				i++
```

Comment:
Do we understand the whole DELETE+INSERT as two different updates? Internally they are, but I think that if we count the UPDATE statements as just one (do we?) then we should do the same thing here. WDYT, @erizocosmico?

Reply:
From looking at the [documentation](), it looks like REPLACE should distinctly act as a shortcut to the two separate calls. The returned count should mimic this behavior:

> The REPLACE statement returns a count to indicate the number of rows affected. This is the sum of the rows deleted and inserted. If the count is 1 for a single-row REPLACE, a row was inserted and no rows were deleted.

I wish I could find it again (my Googling has failed me), but I also remember reading about REPLACE causing a user problems because it was triggering a CASCADE DELETE and messing up a lot more of their data, and the recommendations to that thread was to use UPDATE instead since it's not doing two operations. Totally anecdotal, but it also seems to suggest that this should be treated explicitly as the two separate operations.

Reply:
Skipped that part of the documentation, you are absolutely right. Thank you!